cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/RemoteAppStateStoreTest.java

```diff
@@ -0,0 +1,224 @@
+    cConf.set("app.state.retry.policy.max.retries", "2147483647");
+    cConf.set("app.state.retry.policy.max.time.secs", "60");
+    cConf.set("app.state.retry.policy.type", "exponential.backoff");
+    InMemoryDiscoveryService discoveryService = new InMemoryDiscoveryService();
+    NamespaceAdmin namespaceAdmin = new InMemoryNamespaceAdmin();
+    applicationLifecycleService = Mockito.mock(ApplicationLifecycleService.class);
+    NamespaceMeta testNameSpace = new NamespaceMeta.Builder()
+      .setName(NAMESPACE)
+      .setDescription("This is the default namespace, which is automatically created, and is always available.")
+      .build();
+    namespaceAdmin.create(testNameSpace);
+    remoteClientFactory = new RemoteClientFactory(discoveryService,
+                                                  new DefaultInternalAuthenticator(new AuthenticationTestContext()));
+    httpService = new CommonNettyHttpServiceBuilder(cConf, "appfabric", new NoOpMetricsCollectionService())
+      .setHttpHandlers(new AppStateHandler(applicationLifecycleService, namespaceAdmin)).build();
+    httpService.start();
+    cancellable = discoveryService
+      .register(URIScheme.createDiscoverable(Constants.Service.APP_FABRIC_HTTP, httpService));
+  }
+
+  @AfterClass
+  public static void cleanUp() throws Exception {
+    cancellable.cancel();
+    httpService.stop();
+  }
+
+  @Test
+  public void testSaveSuccess() {
+    RemoteAppStateStore remoteAppStateStore = new RemoteAppStateStore(cConf, remoteClientFactory, NAMESPACE, "app1");
+    byte[] value = "testvalue".getBytes();
+    try {
+      remoteAppStateStore.saveState("key1", value);
+    } catch (IOException e) {
+      //expected
+      Assert.fail("Unexpected exception");
+    }
+  }
+
+  @Test
+  public void testSaveInvalidNamespace() {
+    RemoteAppStateStore remoteAppStateStore = new RemoteAppStateStore(cConf, remoteClientFactory, "invalid", "app1");
+    byte[] value = "testvalue".getBytes();
+    try {
+      remoteAppStateStore.saveState("key1", value);
+      Assert.fail("Expecting an exception");
+    } catch (IOException e) {
+      Assert.assertTrue(e.getCause() instanceof NotFoundException);
+    }
+  }
+
+  @Test
+  public void testSaveInvalidApp() throws ApplicationNotFoundException {
+    String testAppName = "app1";
+    RemoteAppStateStore remoteAppStateStore = new RemoteAppStateStore(cConf, remoteClientFactory, NAMESPACE,
+                                                                      testAppName);
+    byte[] value = "testvalue".getBytes();
+    String testKey = "key1";
+    try {
+
+      Mockito.doThrow(new ApplicationNotFoundException(new ApplicationId(
+        NAMESPACE, testAppName))).when(applicationLifecycleService).saveState(Mockito.any());
+      remoteAppStateStore.saveState(testKey, value);
+      Assert.fail("Expecting an exception");
+    } catch (IOException e) {
+      Assert.assertTrue(e.getCause() instanceof NotFoundException);
```

Comment:
would be a bit cleaner to use the `@Test(expected = NotFoundException.class)` syntax, as it removes the need for the try-catch

Reply:
similar comment for the other tests

Reply:
Since the cause is being checked , using a ExpectedException rule to avoid try/catch.